### PR TITLE
Support cluster-wide crypto policy

### PR DESCRIPTION
**Pull request by borod108**
All webhook servers will now follow the TLS configuration defined in the
SSP CR.
    
This addresses: https://github.com/kubevirt/ssp-operator/issues/407

Co-authored-by: Felix Matouschek <fmatouschek@redhat.com>
Signed-off-by: borod108 <boris.od@gmail.com>
```release-note
SSP-Operator now supports TLS configuration through the SSP CR
```

### CODE REVIEW

File: internal/common/crypto_policy.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+type SSPTLSOptions struct {
+	MinTLSVersion      string
+	OpenSSLCipherNames []string
+}
+
+func (s *SSPTLSOptions) IsEmpty() bool {
+	return len(s.OpenSSLCipherNames) == 0 && s.MinTLSVersion == ""
+}
+
+func (s *SSPTLSOptions) MinTLSVersionId() (minTlsVersionId uint16) {
```
Comment (0xFelix) [inline internal/common/crypto_policy.go:26]: The named return is not used and could be simplified to just the type that is returned.

File: internal/common/crypto_policy.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	return
+}
+
+func emptyTLSProfile() *SSPTLSOptions {
+	return &SSPTLSOptions{"", []string{}}
+}
+
+var (
+	setupLog = ctrl.Log.WithName("setup")
```
Comment (0xFelix) [inline internal/common/crypto_policy.go:70]: If `setupLog` is only used in `GetCiperSuitesAndMinTLSVersion` we should declare it just in this function.

File: internal/common/crypto_policy.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	}
+
+	namespace, err := GetOperatorNamespace(setupLog)
+	if err != nil {
+		return emptyTLSProfile(), err
+	}
+
+	var sspList v1beta1.SSPList
+	if err := apiReader.List(context.TODO(), &sspList, &client.ListOptions{Namespace: namespace}); err != nil {
+		return emptyTLSProfile(), fmt.Errorf("error getCipherSuitesAndMinTLSVersion: %w", err)
```
Comment (0xFelix) [inline internal/common/crypto_policy.go:87]: Do we need to wrap the error? Does it have a benefit?

File: main.go [hunk 4]
```diff
@@ -25,18 +26,20 @@ import (
 
 	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
 	// to ensure that exec-entrypoint and run can make use of them.
-	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	//"k8s.io/apimachinery/pkg/runtime/schema"
```
Comment (0xFelix) [inline main.go:29]: Commented imports?

File: main.go [hunk 5]
```diff
@@ -82,6 +94,20 @@ func runPrometheusServer(metricsAddr string) {
 	}()
 }
 
+func getWebhookServer(sspTLSOptions common.SSPTLSOptions) *webhook.Server {
+	if sspTLSOptions.IsEmpty() {
```
Comment (0xFelix) [inline main.go:98]: What is the default behavior of other parts of SSP when there is no TLS configuration set? Should it run without TLS or not at all? At least all endpoints should behave the same. IIUC the template validator and the prometheus endpoint start with empty TLS config values while the SSP webhook server will refuse to run?

File: internal/common/crypto_policy.go (hunk 3 above)
Comment (borod108) [reply internal/common/crypto_policy.go:87]: I would like to because it is a public method, what do you think?

File: main.go (hunk 4 above)
Comment (borod108) [reply main.go:29]: Yes they are, tnx :P

File: internal/common/crypto_policy.go (hunk 2 above)
Comment (borod108) [reply internal/common/crypto_policy.go:70]: I like to keep the log separate cause the next developer adding a log in some other method will find it easy, even if we use it only there currently, does it make sense? if you find it is not worth it - I will move it.

File: internal/common/crypto_policy.go (hunk 1 above)
Comment (borod108) [reply internal/common/crypto_policy.go]: right!

File: main.go (hunk 5 above)
Comment (borod108) [reply main.go:93]: we decided it will all run if the tls config is not provided. If the webhookserver is set to nil the manager will crate a default one.

File: internal/common/crypto_policy.go (hunk 3 above)
Comment (akrejcir) [reply internal/common/crypto_policy.go:87]: Wrapping error is ok, but the text should provide more information about the function that failed. In this case something like: `"failed listing SSP resources: %w"`.
The calling function of `GetCipherSuitesAndMinTLSVersion()` will know that the failure was in that function, so text `"error getCipherSuitesAndMinTLSVersion"` is useless.

File: internal/common/crypto_policy.go (hunk 2 above)
Comment (0xFelix) [reply internal/common/crypto_policy.go:70]: I would stick to declaring it only in the function because this makes it a "global" variable to the whole `common` package and it is not used anywhere else.

File: internal/common/crypto_policy.go (hunk 3 above)
Comment (0xFelix) [reply internal/common/crypto_policy.go:87]: Ok, but then the other errors which can be returned should be wrapped too with a more explanative error message. You agree?

File: main.go (hunk 5 above)
Comment (0xFelix) [reply main.go:93]: Ah, I forgot that an `nil` webhookserver creates a default one. So everything starts with some default TLS configuration values? That's fine.

File: internal/common/crypto_policy.go (hunk 3 above)
Comment (akrejcir) [reply internal/common/crypto_policy.go:87]: Yes. If we want to wrap errors, then it makes sens to wrap all of them. Unless it is obvious where the error happened without wrapping. But in this case there are multiple calls to the `apiReader` and they could fail with simillar error messages, so wrapping is useful.

File: main.go [hunk 6]
```diff
@@ -17,26 +17,24 @@ limitations under the License.
...
+	"kubevirt.io/ssp-operator/webhooks"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/healthz"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 	"sigs.k8s.io/controller-runtime/pkg/metrics"
-
-	"kubevirt.io/ssp-operator/controllers"
-	"kubevirt.io/ssp-operator/internal/common"
-	"kubevirt.io/ssp-operator/webhooks"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
```
[the remark this answers is not in the document]
Comment (akrejcir) [reply main.go]: If this happens when you edit the text or when you save, then it is an IDE setting. If it happens during `make generate`, then you probably removed an empty line separating groups of imports.

File: internal/operands/template-validator/reconcile.go [hunk 7]
```diff
@@ -155,7 +155,13 @@ func reconcileDeployment(request *common.Request) (common.ReconcileResult, error
 	if request.IsSingleReplicaTopologyMode() && (numberOfReplicas > 1) {
 		numberOfReplicas = 1
 	}
-	deployment := newDeployment(request.Namespace, numberOfReplicas, image)
+
+	sspTLSOptions, err := common.NewSSPTLSOptions(request.Instance.Spec.TLSSecurityProfile, nil)
+	if err != nil {
+		return common.ReconcileResult{}, nil
```
Comment (akrejcir) [inline internal/operands/template-validator/reconcile.go:161]: Can this return the error?

File: internal/common/crypto_policy.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+func NewSSPTLSOptions(tlsSecurityProfile *ocpv1.TLSSecurityProfile, logger *logr.Logger) (*SSPTLSOptions, error) {
+	ciphers, tlsProfile := selectCipherSuitesAndMinTLSVersion(tlsSecurityProfile)
+
+	if logger != nil {
+		logger.Info("Got Ciphers and tlsProfile:", "ciphers: ", ciphers, "tlsProfile: ", tlsProfile)
+	}
+
+	minVersion, err := tlsVersionToHumanReadable(tlsProfile)
+	if err != nil {
+		return nil, nil
```
Comment (akrejcir) [inline internal/common/crypto_policy.go:51]: Can this return the error?

File: internal/common/crypto_policy.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+	if len(sspList.Items) == 0 {
+		return nil, nil
+	}
+
+	ssp := sspList.Items[0]
+
+	sspTLSOptions, err := NewSSPTLSOptions(ssp.Spec.TLSSecurityProfile, &setupLog)
+	if err != nil {
+		return nil, nil
```
Comment (akrejcir) [inline internal/common/crypto_policy.go:97]: Can this return error?

File: internal/common/crypto_policy.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+	case ocpv1.VersionTLS10:
+		return "1.0", nil
+	case ocpv1.VersionTLS11:
+		return "1.1", nil
+	case ocpv1.VersionTLS12:
+		return "1.2", nil
+	case ocpv1.VersionTLS13:
+		return "1.3", nil
+	default:
+		return "0", fmt.Errorf("invalid ocpv1.VersionTLS %v", version)
```
Comment (akrejcir) [inline internal/common/crypto_policy.go:126]: The first return value should be `""`, to be consistent.

File: internal/operands/template-validator/reconcile_test.go [hunk 11]
```diff
@@ -27,6 +27,7 @@ import (
 )
 
 var log = logf.Log.WithName("validator_operand")
+var emptySSPTLSConfig, err = common.NewSSPTLSOptions(nil, nil)
```
Comment (akrejcir) [inline internal/operands/template-validator/reconcile_test.go:30]: Instead of calling the function, you can directly create an empty object:
```go
var emptySSPTLSConfig = &common.SSPTLSOptions{}
```

File: internal/template-validator/tlsinfo/tlsinfo.go [hunk 12]
```diff
@@ -171,4 +188,15 @@ func (ti *TLSInfo) CrateTlsConfig() *tls.Config {
 			return cert, nil
 		},
 	}
+
+	if !ti.sspTLSOptions.IsEmpty() {
+		tlsConfig.CipherSuites = ti.sspTLSOptions.CipherIDs()
+		minVersion, err := ti.sspTLSOptions.MinTLSVersionId()
+		if err != nil {
+			os.Exit(1)
```
Comment (akrejcir) [inline internal/template-validator/tlsinfo/tlsinfo.go:196]: Can you return the error instead?

File: go.mod [hunk 13]
```diff
@@ -10,8 +10,9 @@ require (
 	github.com/go-logr/logr v1.2.3
 	github.com/onsi/ginkgo v1.16.5
 	github.com/onsi/gomega v1.19.0
-	github.com/openshift/api v0.0.0-20220124143425-d74727069f6f // release-4.10
+	github.com/openshift/api v0.0.0-20220525145417-ee5b62754c68 // release-4.10
```
Comment (akrejcir) [inline go.mod:13]: This commit points to the `master` branch. Can you revert this change?

File: go.mod [hunk 14]
```diff
@@ -87,7 +88,7 @@ require (
 	go.uber.org/zap v1.19.1 // indirect
 	golang.org/x/crypto v0.0.0-20220214200702-86341886e292 // indirect
 	golang.org/x/net v0.0.0-20220225172249-27dd8689420f // indirect
-	golang.org/x/oauth2 v0.0.0-20220223155221-ee480838109b // indirect
+	golang.org/x/oauth2 v0.0.0-20220411215720-9780585627b5 // indirect
```
Comment (akrejcir) [inline go.mod:91]: Is this change needed?

**Review by akrejcir — COMMENTED** [review]
/lgtm

**Comment by akrejcir** [comment]
/ltgm

Please remove the `WIP` from title.
